### PR TITLE
Add UUID suffix support

### PR DESCRIPTION
**Pull request by dbarbuzzi**
Add support for UUID-based suffixes (UUID4 or UUID7). The default is still timestamp (open to discussion). There is one very light dependency added because UUID7 is not available in the Python stdlib.

### CODE REVIEW

File: src/pytest_nm_releng/plugin.py [hunk 1]
```diff
@@ -15,18 +15,45 @@
...
+    }
+    return suffix_map.get(suffix_type, get_utc_timestamp)()
 
 
 def generate_junit_flags() -> list[str]:
     if not (junitxml_base_dir := os.getenv("NMRE_JUNIT_BASE")):
         return []
 
-    junitxml_file = Path(junitxml_base_dir) / f"{get_utc_timestamp()}.xml"
+    junit_suffix_type = os.getenv("NMRE_JUNIT_SUFFIX_TYPE")
```
Comment (derekk-nm) [inline src/pytest_nm_releng/plugin.py]: why not just set the default `SuffixType.TIMESTAMP` in this method call?

**Review by derekk-nm — DISMISSED** [review]
minor question, but lgtm.  thank you!

File: src/pytest_nm_releng/plugin.py (hunk 1 above)
Comment (dbarbuzzi) [reply src/pytest_nm_releng/plugin.py]: Good question, I’ll take a closer look at that.

Comment (dbarbuzzi) [reply src/pytest_nm_releng/plugin.py]: I have updated this area to set the default to be a little extra explicit (and defined the default separately to make any changes to it cleaner), but this actually doesn't reduce any logic below because having the default only helps if the environment variable is unset, not if it is e.g., defined but empty or set to an invalid value.

I suppose there is an argument to be made about invalid values causing errors, but here I was leaning toward just replacing with the default value (maybe emit a warning?). If you feel differently, let me know (also if you consider an empty value to be in the same vein as other invalid values).

File: tests/test_helpers.py [hunk 2]
```diff
@@ -18,33 +18,52 @@
...
 
-from pytest_nm_releng.plugin import generate_junit_flags
+from pytest_nm_releng.plugin import DEFAULT_SUFFIX_TYPE, generate_junit_flags
 from tests.utils import setenv
 
 EnvVarValue = Union[str, None]
 
+SUFFIX_PATTERN_MAP = {
+    "timestamp": r"\d{10,}\.\d+\.xml",
+    "uuid4": r"[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\.xml",
```
Comment (dhuangnm) [inline tests/test_helpers.py:28]: What is the difference between uuid4 and uuid7? From the output file name format, they look the same?

File: src/pytest_nm_releng/plugin.py [hunk 3]
```diff
@@ -15,18 +15,48 @@
...
+        if junit_suffix_type in SuffixType._value2member_map_
+        else SuffixType.TIMESTAMP
+    )
+
+    junitxml_file = (
+        Path(junitxml_base_dir) / f"{generate_suffix(junit_suffix_type)}.xml"
+    )
 
     if prefix := os.getenv("NMRE_JUNIT_PREFIX"):
         junitxml_file = junitxml_file.with_name(f"{prefix}{junitxml_file.name}")
```
Comment (dhuangnm) [inline src/pytest_nm_releng/plugin.py]: Should we apply "Path(junitxml_base_dir) /" as a last step after prefix is applied first?

File: tests/test_helpers.py [hunk 4]
```diff
@@ -18,33 +18,52 @@
...
-        pytest.param("test-results", "run-", id="base-with-prefix"),
-        pytest.param("test-results", "", id="base-with-empty-prefix"),
-        pytest.param("test-results", None, id="base-with-unset-prefix"),
-        pytest.param("", "run-", id="empty-base-with-prefix"),
-        pytest.param("", "", id="empty-base-with-empty-prefix"),
-        pytest.param("", None, id="empty-base-with-unset-prefix"),
-        pytest.param(None, "run-", id="unset-base-with-prefix"),
-        pytest.param(None, "", id="unset-base-with-empty-prefix"),
-        pytest.param(None, None, id="unset-base-with-unset-prefix"),
+        # base path present
```
Comment (dhuangnm) [inline tests/test_helpers.py:37]: What do these pytest.param() do? It seems there are more combinations of base, prefix and suffix?

File: tests/test_helpers.py (hunk 2 above)
Comment (dbarbuzzi) [reply tests/test_helpers.py:28]: There are several versions/variants of UUIDs that all result in a 32-character hexadecimal string (or 36 when formatted in the common 8-4-4-4-12 structure including dashes). You could tell them apart from the 13th hex character (first character in the third block) – this will be a `4` for `uuid4` and a `7` for `uuid7`.

The more common variants are UUID4 and UUID7. The gist of these two are:
* `uuid4` is a general-purpose randomly generated identifier
* `uuid7` is similar, but the first several characters encode the time of creation and it is often used for distributed purposes (which I think our use case would technically qualify as we 'distribute' multiple test/etc. runs that would use the UUIDs over multiple jobs/runners).

I will leave further research of the variants to the reader :)

File: src/pytest_nm_releng/plugin.py (hunk 3 above)
Comment (dbarbuzzi) [reply src/pytest_nm_releng/plugin.py]: I like having the prefix after mainly to flow through the feature logic in a particular way (the prefix is an optional 'input' so we conditionally apply it after the pieces that are required). However, I can see the logic in moving it before the path creation and I think that would also eliminate a single line of code. I’ll think about this a bit and probably adopt the change.

File: tests/test_helpers.py (hunk 4 above)
Comment (dbarbuzzi) [reply tests/test_helpers.py:37]: All the params are providing different combinations of 'inputs' that the plugin might receive to run several different input scenarios against a singular test.

`pytest.param` specifically allows you to wrap individual parametrizations so you can provide additional data, namely the `id` being provided here, which allows better description of the test cases in the `pytest` output/reporting.

For example, in this test case, a `None` input for one of the environment variables means that environment variable will be unset for the test, so using the `id` kwarg allows better annotation of that detail.

All that said, there are more variations for base & prefix because those are more 'variable' than the suffix, and control more. If the base is unset or empty, it fully disables this functionality. If prefix is unset/empty, it shouldn't have a prefix, but the base should still work.

By contrast, the suffix is more specific: it is tested for being empty/unset (in which case it uses the default), or it allows a specific set of values which are each individually checked that they produce the corresponding suffix.

That said, I realize I forgot to add cases where the suffix is set to an invalid value, so I will add a separate test case for that with a few scenarios/parametrizations of invalid values.

File: tests/test_helpers.py (hunk 2 above)
Comment (dhuangnm) [reply tests/test_helpers.py:28]: I see, thanks for the information.

File: src/pytest_nm_releng/plugin.py (hunk 3 above)
Comment (dhuangnm) [reply src/pytest_nm_releng/plugin.py]: The main issue was that the prefix was applied before the base/suffix.xml but I see the updated code looks good now, thanks.

Comment (dbarbuzzi) [reply src/pytest_nm_releng/plugin.py]: Ah, yeah – the prefix has been an existing feature for a while and the way it was applied prior was not an issue with the way the `Path` methods work – basically, it was replacing the existing filename (just the filename and not any folder names before it) with the new filename that was constructed with the prefix + original filename.

File: src/pytest_nm_releng/plugin.py (hunk 1 above)
Comment (derekk-nm) [reply src/pytest_nm_releng/plugin.py]: i'm good.  thanks for the explaination

**Review by derekk-nm — APPROVED** [review]

File: tests/test_helpers.py (hunk 4 above)
Comment (dhuangnm) [reply tests/test_helpers.py:37]: I see, thanks for the explanation.

**Review by dhuangnm — APPROVED** [review]
LGTM, it'll be good to run it with some test jobs first before merging to verify the output files are generated as expected.

**Comment by dbarbuzzi** [comment]
> LGTM, it'll be good to run it with some test jobs first before merging to verify the output files are generated as expected.

This is actually covered by the unit tests.

**edit** – but yes, any jobs that update the version (once I tag a new one) will of course be tested before being merged.
